### PR TITLE
Add FromVariant macro and cleanup codebase

### PR DESCRIPTION
**Pull request by i1i1**
This derive macro is used for making `From<Variant>` trait for enums.

Signed-off-by: i1i1 <vanyarybin1@live.ru>

### CODE REVIEW

**Review by humb1t — APPROVED** [review]

File: iroha_macro/iroha_derive/src/lib.rs [hunk 1]
```diff
@@ -155,6 +161,93 @@ fn impl_io(ast: &syn::DeriveInput) -> TokenStream {
     gen.into()
 }
 
+fn get_struct_arg<'a, 'b>(s: &'a str, ty: &'b syn::TypePath) -> Option<&'b syn::GenericArgument> {
+    let segs = &ty.path.segments;
```
Comment (eadventurous) [inline iroha_macro/iroha_derive/src/lib.rs:165]:
```suggestion
    let segments = &ty.path.segments;
```

File: iroha_macro/iroha_derive/src/lib.rs [hunk 2]
```diff
@@ -155,6 +161,93 @@ fn impl_io(ast: &syn::DeriveInput) -> TokenStream {
...
+    if segs.len() != 1 || segs[0].ident != s {
+        return None;
+    }
+
+    if let syn::PathArguments::AngleBracketed(ref bracargs) = segs[0].arguments {
```
Comment (eadventurous) [inline iroha_macro/iroha_derive/src/lib.rs:170]:
```suggestion
    if let syn::PathArguments::AngleBracketed(ref bracketed_arguments) = segs[0].arguments {
```

File: iroha_macro/iroha_derive/src/lib.rs [hunk 3]
```diff
@@ -155,6 +161,93 @@ fn impl_io(ast: &syn::DeriveInput) -> TokenStream {
     gen.into()
 }
 
+fn get_struct_arg<'a, 'b>(s: &'a str, ty: &'b syn::TypePath) -> Option<&'b syn::GenericArgument> {
```
Comment (eadventurous) [inline iroha_macro/iroha_derive/src/lib.rs:164]:
```suggestion
fn get_struct_argument<'a, 'b>(s: &'a str, ty: &'b syn::TypePath) -> Option<&'b syn::GenericArgument> {
```

File: iroha_macro/iroha_derive/src/lib.rs [hunk 4]
```diff
@@ -155,6 +161,93 @@ fn impl_io(ast: &syn::DeriveInput) -> TokenStream {
...
+    from_orig
+}
+
+fn impl_from_variant(ast: &syn::DeriveInput) -> TokenStream {
+    let name = &ast.ident;
+
+    let froms = if let syn::Data::Enum(ref data_enum) = ast.data {
+        &data_enum.variants
+    } else {
+        unimplemented!()
```
Comment (eadventurous) [inline iroha_macro/iroha_derive/src/lib.rs:232]: I would suggest to give some context, e.g. show a user `Only supports enums` or something like that.

File: iroha_macro/iroha_derive/src/lib.rs [hunk 5]
```diff
@@ -155,6 +161,93 @@ fn impl_io(ast: &syn::DeriveInput) -> TokenStream {
...
+    }
+    .iter()
+    .filter_map(|variant| {
+        if let syn::Fields::Unnamed(ref unnamed) = variant.fields {
+            if unnamed.unnamed.len() == 1 {
+                return Some((&variant.ident, &unnamed.unnamed.first().unwrap().ty));
```
Comment (eadventurous) [inline iroha_macro/iroha_derive/src/lib.rs:238]:
```suggestion
                return Some((&variant.ident, &unnamed.unnamed.first().expect(...).ty));
```

File: iroha_macro/iroha_derive/src/lib.rs [hunk 6]
```diff
@@ -155,6 +161,93 @@ fn impl_io(ast: &syn::DeriveInput) -> TokenStream {
...
+
+fn from_variant(
+    into_ty: &syn::Ident,
+    from_variant: &syn::Ident,
+    from_ty: &syn::Type,
+) -> proc_macro2::TokenStream {
+    let from_orig = from_variant_internal(into_ty, from_variant, from_ty);
+
+    if let syn::Type::Path(path) = from_ty {
+        if let Some(inner) = get_struct_arg("Box", path) {
```
Comment (eadventurous) [inline iroha_macro/iroha_derive/src/lib.rs:214]: Better to put raw string values into constants.

File: iroha_macro/iroha_derive/src/lib.rs [hunk 7]
```diff
@@ -155,6 +161,93 @@ fn impl_io(ast: &syn::DeriveInput) -> TokenStream {
...
+    let segs = &ty.path.segments;
+    if segs.len() != 1 || segs[0].ident != s {
+        return None;
+    }
+
+    if let syn::PathArguments::AngleBracketed(ref bracargs) = segs[0].arguments {
+        assert_eq!(bracargs.args.len(), 1);
+        Some(&bracargs.args[0])
+    } else {
+        unimplemented!()
```
Comment (eadventurous) [inline iroha_macro/iroha_derive/src/lib.rs:174]: I think it is better to panic with an explanation message.

**Review by eadventurous — CHANGES_REQUESTED** [review]
Please use full words in macro related functions, it is very difficult to read now.

**Review by eadventurous — APPROVED** [review]

**Pull request merged**